src/ATokenCheckpointed.sol

```diff
@@ -0,0 +1,307 @@
+    internal
+    returns (uint256 _previousBalance, uint256 _currentBalance)
+  {
+    (_previousBalance, _currentBalance) = balanceCheckpoints[_user].push(_rawBalanceOf(_user));
+  }
+
+  // Returns the raw (i.e. unrebased) balanceOf the _user at the _blockNumber.
+  function getPastStoredBalance(address _user, uint256 _blockNumber) public returns (uint256) {
+    // TODO could we use getAtProbablyRecentBlock instead?
+    return balanceCheckpoints[_user].getAtBlock(_blockNumber);
+  }
+
+  function getPastTotalDeposits(uint256 _blockNumber) public returns (uint256) {
+    return totalDepositCheckpoints.getAtBlock(_blockNumber);
+  }
+
+  // forgefmt: disable-start
+  //===========================================================================
+  // BEGIN: Aave overrides
+  //===========================================================================
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaledWithCheckpoint function.
+  ///
+  /// @inheritdoc IAToken
+  function mint(
+    address caller,
+    address onBehalfOf,
+    uint256 amount,
+    uint256 index
+  ) external virtual override onlyPool returns (bool) {
+    return _mintScaledWithCheckpoint(caller, onBehalfOf, amount, index);
+  }
+
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaledWithCheckpoint function.
+  ///
+  /// @inheritdoc IAToken
+  function mintToTreasury(uint256 amount, uint256 index) external override onlyPool {
+    if (amount == 0) {
+      return;
+    }
+    _mintScaledWithCheckpoint(address(POOL), _treasury, amount, index);
+  }
+
+  /// Note: this has been modified from Aave v3's AToken to checkpoint raw
+  /// balances. We cannot just call `super` here because the function is external.
+  ///
+  /// @inheritdoc IAToken
+  function burn(
+    address from,
+    address receiverOfUnderlying,
+    uint256 amount,
+    uint256 index
+  ) external virtual override onlyPool {
+    _burnScaled(from, receiverOfUnderlying, amount, index);
+
+    // Begin modifications.
+    (uint256 _preBurnRawBalance, uint256 _postBurnRawBalance) = _checkpointRawBalanceOf(from);
+    totalDepositCheckpoints.push(
+      totalDepositCheckpoints.latest() - (_preBurnRawBalance - _postBurnRawBalance)
+    );
+    // End modifications.
+
+    if (receiverOfUnderlying != address(this)) {
+      IERC20(_underlyingAsset).safeTransfer(receiverOfUnderlying, amount);
+    }
+  }
+
+  /// Note: this has been modified from Aave v3's IncentivizedERC20 contract to
+  /// checkpoint raw balances accordingly.
+  ///
+  /// @inheritdoc IERC20
+  function transfer(address recipient, uint256 amount) external virtual override returns (bool) {
+    uint128 castAmount = amount.toUint128();
+    _transfer(_msgSender(), recipient, castAmount);
+    // Begin modifications.
+    _checkpointRawBalanceOf(_msgSender());
+    _checkpointRawBalanceOf(recipient);
+    // End modifications.
```

Comment:
^^ the internal `_transfer` instead

agreed

it seems like there are quite a few internal `_transfer`s! We should be sure to override the right one (aka should checkpoint on both `AToken:_transfer` fns, the one with a validate bool and the one without)

Reply:
Ahh yeah, great catch, guys!

Reply:
> We should be sure to override the right one (aka should checkpoint on both AToken:_transfer fns, the one with a validate bool and the one without)

I don't think we need to checkpoint in both, since the one without the validation argument [just calls the one with it](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol#L239). If we just checkpoint in the latter, that should cover us. But I agree we need to make sure we're overriding the correct function.

Unfortunately I think [the function we want to override](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol#L206-L211) isn't virtual. This is another one that we're going to need to get changed upstream. I'll just submit a PR at this point.

Reply:
This fixes the problem locally (where I made `_transfer/4` virtual): fccbe61

Reply:
>  I'll just submit a PR at this point.

Cool, yea IMO we should prioritize this and then simplify our code. Is your current thinking to keep the `_transfer` override and leave this PR open until the upstream PR is merged, or to revert this `_transfer` change (since it doesn't compile) then handle that in a separate PR? Guess it depends how quick they can merge it upstream

Reply:
Aave PR is here: https://github.com/aave/aave-v3-core/pull/774

Reply:
Aave merged our PR earlier today and I've bumped the dependency in [e3ab13b](https://github.com/ScopeLift/flexible-voting/pull/21/commits/e3ab13b2366ec9238d037d49d2e5016502db31f2). So this should be taken care of now